[Merged by Bors] - Move generating TXs in tests into `svm`; add & use spawn TX in tests

nkryuchkov opened the pull request:
## Motivation
<!-- Please mention the issue fixed by this PR or detailed motivation -->
Closes #2829
<!-- `Closes #XXXX, closes #XXXX, ...` links mentioned issues to this PR and automatically closes them when this it's merged -->

## Changes
<!-- Please describe in detail the changes made -->
- Move code for generating transactions into `svm` package
- Add spawn transactions and use them in tests

## Test Plan
<!-- Please specify how these changes were tested 
(e.g. unit tests, manual testing, etc.) -->

## TODO
<!-- This section should be removed when all items are complete -->
- [ ] Explain motivation or link existing issue(s)
- [ ] Test changes and document test plan
- [ ] Update documentation as needed

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

nkryuchkov committed:
Change way of generating transactions and addresses in tests

nkryuchkov committed:
Fix algorithm of GenerateAddress

nkryuchkov committed:
Merge branch 'develop' into feature/move-tests-tx-gen-into-svm

# Conflicts:
#	mesh/mesh_test.go
#	mesh/meshdb_test.go
#	mesh/reward_test.go

nkryuchkov committed:
Add comment to GenerateAddress

nkryuchkov committed:
Fix compilation error after merging

nkryuchkov committed:
Change [20] to [types.AddressLength]

nkryuchkov committed:
Fix a comment

noamnelke commented on common/types/address.go:
```diff
@@ -12,10 +17,11 @@ import (
 )
 
-// Address represents the 20 byte address of an spacemesh account.
+// Address represents the address of an spacemesh account with AddressLength length.
```
```suggestion
// Address represents the address of a spacemesh account with AddressLength length.
```

countvonzero commented on activation/atxdb_test.go:
```diff
@@ -739,12 +739,12 @@ func TestActivationDb_ContextuallyValidateAtx(t *testing.T) {
 	atxdb := NewDB(database.NewMemDatabase(), nil, idStore, layersPerEpochBig, goldenATXID, &ValidatorMock{}, lg.WithName("atxDB"))
 
-	validAtx := types.NewActivationTx(newChallenge(nodeID, 0, *types.EmptyATXID, goldenATXID, types.LayerID{}), [20]byte{}, nil, 0, nil)
+	validAtx := types.NewActivationTx(newChallenge(nodeID, 0, *types.EmptyATXID, goldenATXID, types.LayerID{}), [types.AddressLength]byte{}, nil, 0, nil)
```
just `types.Address{}`, here and below

countvonzero commented on common/types/address.go:
```diff
@@ -12,10 +17,11 @@ import (
 
 const (
 	// AddressLength is the expected length of the address.
+	// TODO(nkryuchkov): change to 32.
```
why? can you put the issue number here? thanks

nkryuchkov replied:
I assumed it should be changed to 32, but I just thought about it and now I'm not sure it should be. @noamnelke @neysofu will it be 20 or 32?

noamnelke replied:
it should be 20 :)

nkryuchkov replied:
My bad. Removed

countvonzero commented on common/types/address.go:
```diff
@@ -112,3 +118,43 @@ func (a *Address) SetBytes(b []byte) {
+
+type nonce128 struct {
+	mu          sync.Mutex
+	little, big uint64
```
is it possible to encode this with xdr in the codec package?

nkryuchkov replied:
Could you explain your thought, please? I didn't understand

nkryuchkov replied:
https://github.com/spacemeshos/go-spacemesh/pull/3046#discussion_r781388828
@countvonzero I'm taking my last words back 🙂

countvonzero commented on common/types/address.go:
```diff
@@ -112,3 +118,43 @@ func (a *Address) SetBytes(b []byte) {
+	return a
+}
+
+var nonce nonce128
```
this implies all addresses will share the same nonce?

nkryuchkov replied:
No, for now, all nonces are just zeros of size 128 bit. Later, we will need to have a nonce implementation. Perhaps I need to simplify the code to make it more clear

noamnelke replied:
See my comment below, I think this should be removed.

nkryuchkov replied:
I'll remove it according to https://github.com/spacemeshos/go-spacemesh/pull/3046#discussion_r781388327

countvonzero commented on common/types/address.go:
```diff
@@ -112,3 +118,43 @@ func (a *Address) SetBytes(b []byte) {
 	copy(a[AddressLength-len(b):], b)
 }
+
+type nonce128 struct {
```
i think we need some comment on the nonce here. (what is little? big?)

noamnelke replied:
I don't think this whole struct is not needed. See my comment on where it's used.

nkryuchkov replied:
https://github.com/spacemeshos/go-spacemesh/pull/3046#discussion_r781388828

countvonzero commented on common/types/address.go:
```diff
@@ -112,3 +118,43 @@ func (a *Address) SetBytes(b []byte) {
+
+func (n *nonce128) next() []byte {
+	n.mu.Lock()
+	defer n.mu.Unlock()
```
why not instantiate one nonce per address? i don't understand the need for mutex

nkryuchkov replied:
https://github.com/spacemeshos/go-spacemesh/pull/3046#discussion_r781388828

countvonzero commented on common/types/transaction.go:
```diff
@@ -192,8 +189,13 @@ type Reward struct {
-// NewSignedTx is used in TESTS ONLY to generate signed txs.
-func NewSignedTx(nonce uint64, rec Address, amount, gas, fee uint64, signer *signing.EdSigner) (*Transaction, error) {
+// GenerateSpawnTransaction generates a spawn transaction.
+func GenerateSpawnTransaction(signer *signing.EdSigner, target Address) *Transaction {
```
unused?

noamnelke replied:
See my review comment - we should start using it in this PR before merging.

countvonzero commented on common/types/transaction.go:
```diff
@@ -192,8 +189,13 @@ type Reward struct {
+}
+
+// GenerateCallTransaction generates a call transaction.
+func GenerateCallTransaction(signer *signing.EdSigner, rec Address, nonce, amount, gas, fee uint64) (*Transaction, error) {
```
if this is intended for testing only, can you move it to common/type/testutil.go?

noamnelke replied:
While this will only be used in tests - it should actually live inside the SVM package (see my review comment).

noamnelke commented on common/types/address.go:
```diff
@@ -112,3 +118,43 @@ func (a *Address) SetBytes(b []byte) {
+// GenerateAddress generates an address from a public key.
+func GenerateAddress(publicKey []byte) Address {
+	input := bytes.Repeat([]byte{0x00}, 20)
+	input = append(input, nonce.next()...)
```
Not sure I understand this... Address generation should be deterministic, so why are we hashing a nonce into it?

Also not sure why we're hashing 20 zero-bytes into the address...

nkryuchkov replied:
As I understood, the generated address should be `SHA256(20 null-bytes || 128-bit-nonce || PubKey)`. If not, how should it be generated?

nkryuchkov replied:
@noamnelke and I had a discussion about this — `GenerateAddress` should return the last 20 bytes of the public key. So I'll change this

noamnelke reviewed (changes requested):
What's missing:
- Some cleanup, e.g. `StringToAddress` from `common/types/address.go`.
- Replacing the various implementations of `NewTx` in tests with the new function.
- Actually use `GenerateSpawnTransaction` in tests. The idea is that when we switch to SVM a transaction from a new account that hasn't been spawned yet will fail. We want to update the tests now so that we won't need to touch them as part of the transition to SVM. In simple tests we can just execute a spawn transaction before the first transaction from an account. In tests where transactions are generated more dynamically (if we have something like that) we can check if the account nonce is zero and then generate a spawn transaction.
- Last, but not least: the code for generating an address and a transaction should live inside the SVM package. This is critical so that when we replace the package those get replaced as well. The `types` package should not need to import those methods, if I'm not mistaken - let me know if I missed something and we'll figure it out.

nkryuchkov committed:
CR feedback part 1

neysofu reviewed (changes requested):
I agree with @noamnelke 's assessment on switching to `GenerateSpawnTransaction` in tests already and defining `GenerateCallTransaction` in the `svm` package. Otherwise LGTM.

nkryuchkov committed:
Use spawn tx

nkryuchkov committed:
Move GenerateSpawnTransaction and GenerateCallTransaction inside `svm` package

nkryuchkov committed:
Remove `SubmitTransaction_ZeroBalance`

nkryuchkov committed:
Fix mesh tests

nkryuchkov committed:
Fix `TestSpacemeshApp_TransactionService`

nkryuchkov committed:
Merge branch 'develop' into feature/move-tests-tx-gen-into-svm

# Conflicts:
#	mesh/meshdb_test.go

nkryuchkov committed:
Fix tests

nkryuchkov committed:
Fix linter

countvonzero commented on activation/atxdb_test.go:
```diff
@@ -741,12 +741,12 @@ func TestActivationDb_ContextuallyValidateAtx(t *testing.T) {
 
 	arbitraryAtxID := types.ATXID(types.HexToHash32("11111"))
-	malformedAtx := types.NewActivationTx(newChallenge(nodeID, 0, arbitraryAtxID, goldenATXID, types.LayerID{}), [20]byte{}, nil, 0, nil)
+	malformedAtx := types.NewActivationTx(newChallenge(nodeID, 0, arbitraryAtxID, goldenATXID, types.LayerID{}), [types.AddressLength]byte{}, nil, 0, nil)
```
:s/[types.AddressLength]byte{}/types.Address{}

countvonzero commented on activation/atxdb_test.go:
```diff
@@ -793,7 +793,7 @@ func BenchmarkGetAtxHeaderWithConcurrentStoreAtx(b *testing.B) {
 			id := types.NodeID{Key: util.Bytes2Hex(pub), VRFPublicKey: []byte("22222")}
 			for i := 0; ; i++ {
-				atx := types.NewActivationTx(newChallenge(id, uint64(i), *types.EmptyATXID, goldenATXID, types.NewLayerID(0)), [20]byte{}, nil, 0, nil)
+				atx := types.NewActivationTx(newChallenge(id, uint64(i), *types.EmptyATXID, goldenATXID, types.NewLayerID(0)), [types.AddressLength]byte{}, nil, 0, nil)
```
:s/[types.AddressLength]byte{}/types.Address{}

countvonzero commented on pendingtxs/pending_txs.go:
```diff
@@ -136,7 +136,7 @@ func validTxWithHighestFee(txs map[types.TransactionID]nanoTx, balance uint64) t
 	var maxFee uint64
 	for id, tx := range txs {
-		if (tx.Fee > maxFee || (tx.Fee == maxFee && bytes.Compare(id[:], bestID[:]) < 0)) &&
+		if (tx.Fee > maxFee || (tx.Fee == maxFee && bytes.Compare(id[:], bestID[:]) < 0) || (tx.Fee == 0 && maxFee == 0)) &&
```
i think this will be clearer to rewrite to
```
if balance >= (tx.Amount+tx.Fee) && 
(bestID == types.EmptyTransactionID || // first tx with enough balance
 tx.Fee > maxFee || 
 (tx.Fee == maxFee && bytes.Compare(id[:], bestID[:]) < 0))
```

nkryuchkov replied:
thanks

countvonzero commented on svm/transaction/transaction.go:
```diff
@@ -0,0 +1,56 @@
+		Signature:        [64]byte{},
+	}
+
+	copy(sst.Signature[:], signer.Sign(buf))
```
why do this extra copy and not directly do
```
sst := &types.Transaction{
	InnerTransaction: inner,
	Signature:        signer.Sign(buf),
}
```

nkryuchkov replied:
@countvonzero I copied it from the existing code. If we assume that `signer.Sign(buf)` always returns a 64-byte result, your suggestion would work the same way.

nkryuchkov replied:
@countvonzero I looked at it and actually `signer.Sign(buf)` returns a slice, but we need an array, so we need to convert it using `copy`

countvonzero commented on trie/trie_test.go:
```diff
@@ -569,7 +569,7 @@ func BenchmarkHash(b *testing.B) {
 
 	// Create a realistic account trie to hash
-	addresses := make([][20]byte, b.N)
+	addresses := make([][types.AddressLength]byte, b.N)
```
addresses := make([]types.Address, b.N)

nkryuchkov committed:
CR feedback

dshulyak commented:
@nkryuchkov is this change about moving code for generating transactions to svm module, any other important changes? can it be reflected in the pr title, in the current state the title is somewhat obscure

countvonzero reviewed (commented):
code LGTM

please update the PR content (currently empty) to detail what was done.

will leave the approval to @neysofu and @noamnelke

nkryuchkov commented:
@dshulyak @countvonzero thanks, updated

nkryuchkov commented:
bors merge

neysofu commented:
Needs approval from @noamnelke

nkryuchkov commented:
bors merge

noamnelke commented on cmd/node/app_test.go:
```diff
@@ -267,7 +268,7 @@ func txWithUnorderedNonceGenerator(dependencies []int) TestScenario {
 	setup := func(suite *AppTestSuite, t *testing.T) {
 		for i := 0; i < txsSent; i++ {
-			tx, err := types.GenerateCallTransaction(acc1Signer, dst, uint64(txsSent-i), 10, 1, 1)
+			tx, err := transaction.GenerateCallTransaction(acc1Signer, dst, uint64(txsSent-i), 10, 1, 1)
```
I haven't followed the entire test here, but if we didn't add a spawn transaction at nonce == 0, won't this test fail when those are needed?

We should probably make sure that tests without a spawn transaction fail.

nkryuchkov replied:
This `txWithUnorderedNonceGenerator` test is intended to error at https://github.com/spacemeshos/go-spacemesh/blob/feature/move-tests-tx-gen-into-svm/cmd/node/app_test.go#L278 because nonces are out of order: from 25 to 0. Perhaps, it would look clearer if the last TX with nonce 0 would be spawn TX, but that wouldn't change the behavior, so IMO it doesn't matter here

noamnelke commented on cmd/node/app_test.go:
```diff
@@ -329,8 +330,14 @@ func txWithRunningNonceGenerator(dependencies []int) TestScenario {
-			suite.NoError(err, "failed to create signed tx: %s", err)
+			var tx *types.Transaction
+			if i == 0 {
+				tx = transaction.GenerateSpawnTransaction(acc1Signer, dst)
```
I'm wondering how this can work... If the spawn transaction is basically empty and it is the only transaction that gets the nonce 0 and accounts are still initialized with nonce == 0 and skipping is not allowed. How can any transaction make it in?

nkryuchkov replied:
What do you mean by "skipping is not allowed"? If the nonce is 0, a spawn TX is generated here, then nonce gets increased inside https://github.com/spacemeshos/go-spacemesh/blob/feature/move-tests-tx-gen-into-svm/cmd/node/app_test.go#L344, so next TXs become call TXs

noamnelke commented on mempool/tx_pool_test.go:
```diff
@@ -173,7 +174,7 @@ func createBatch(t testing.TB, signer *signing.EdSigner) ([]*types.Transaction,
 	var txIDBatch []types.TransactionID
 	for i := uint64(0); i < 10000; i++ {
-		tx, err := types.GenerateCallTransaction(signer, types.Address{}, 5+1, 50, 100, 1)
+		tx, err := transaction.GenerateCallTransaction(signer, types.Address{}, 5+1, 50, 100, 1)
```
No spawn transactions in this file?

nkryuchkov replied:
No, because they are not submitted. There's no nonce == 0 in this file, so as I understand, there's no need for spawn TXs here. Please correct me if I'm wrong

noamnelke commented on mesh/mesh_test.go:
```diff
@@ -76,7 +77,7 @@ func addTransactions(t testing.TB, mesh *DB, layerID types.LayerID) []types.Tran
 	txIDs := make([]types.TransactionID, 0, numTXs)
 	for i := 0; i < numTXs; i++ {
-		tx, err := types.GenerateCallTransaction(signing.NewEdSigner(), types.HexToAddress("1"), 1, 10, 100, rand.Uint64())
+		tx, err := transaction.GenerateCallTransaction(signing.NewEdSigner(), types.HexToAddress("1"), 1, 10, 100, rand.Uint64())
```
Spawn transactions?

nkryuchkov replied:
https://github.com/spacemeshos/go-spacemesh/pull/3046#discussion_r790871476

noamnelke commented on mesh/mesh_test.go:
```diff
@@ -545,7 +546,7 @@ func TestMesh_persistLayerHash(t *testing.T) {
 func addTxToMempool(t *testing.T, msh *Mesh, signer *signing.EdSigner, nonce uint64) *types.Transaction {
 	t.Helper()
-	tx := newTx(t, signer, nonce, 111)
+	tx := newCallTx(t, signer, nonce, 111)
```
Spawn?

nkryuchkov replied:
https://github.com/spacemeshos/go-spacemesh/pull/3046#discussion_r790871476

noamnelke commented on mesh/meshdb_test.go:
```diff
@@ -227,8 +234,8 @@ func TestMeshDB_GetStateProjection_WrongNonce(t *testing.T) {
 	err := mdb.writeTransactions(types.NewLayerID(1), types.BlockID{1},
-		newTx(t, signer, 1, 10),
-		newTx(t, signer, 2, 20),
+		newCallTx(t, signer, 1, 10),
```
Spawn?

noamnelke commented on miner/proposal_builder_test.go:
```diff
@@ -74,7 +75,7 @@ func createBuilder(tb testing.TB) *testBuilder {
 func genTX(tb testing.TB, nonce uint64, recipient types.Address, signer *signing.EdSigner) *types.Transaction {
 	tb.Helper()
-	tx, err := types.GenerateCallTransaction(signer, recipient, nonce, 1, defaultGasLimit, defaultFee)
+	tx, err := transaction.GenerateCallTransaction(signer, recipient, nonce, 1, defaultGasLimit, defaultFee)
```
Spawn?

noamnelke commented on sql/transactions/transactions_test.go:
```diff
@@ -28,9 +29,9 @@ func TestGetHas(t *testing.T) {
-		mustTx(types.NewSignedTx(1, types.Address{1}, 191, 1, 1, signer1)),
-		mustTx(types.NewSignedTx(1, types.Address{2}, 191, 1, 1, signer2)),
-		mustTx(types.NewSignedTx(1, types.Address{3}, 191, 1, 1, signer1)),
+		mustTx(transaction.GenerateCallTransaction(signer1, types.Address{1}, 1, 191, 1, 1)),
```
Aren't spawn transactions needed in this entire file?

noamnelke commented on svm/api_test.go:
```diff
@@ -37,7 +38,7 @@ func (appliedTxsMock) NewBatch() database.Batch           { panic("implement me"
 
 func createTransaction(t *testing.T, nonce uint64, destination types.Address, amount, fee uint64, signer *signing.EdSigner) *types.Transaction {
-	tx, err := types.GenerateCallTransaction(signer, destination, nonce, amount, 100, fee)
+	tx, err := transaction.GenerateCallTransaction(signer, destination, nonce, amount, 100, fee)
```
Spawn?

noamnelke reviewed (commented):
@nkryuchkov it seems that the addition of spawn transactions is inconsistent (sometimes you haven't done it). This made me realize that nothing checks for this. I think that tests should fail when a spawn transaction isn't added.

Additionally, I don't fully understand how it currently works WITH spawn transactions. Do they get processed and increment the account nonce? I couldn't see the code that does that. Otherwise, assuming the account nonce is initialized to zero, and we replace transactions with that nonce to spawn transactions - how does the first call transaction (with nonce == 1) get applied?

nkryuchkov commented:
@noamnelke 

I added spawn TXs everywhere where nonce was 0 in tests. In other cases, TXs were not submitted, so I assumed that nonce was not checked, therefore there was no need for spawn TXs, so I didn't add spawn TXs there.

> Do they get processed and increment the account nonce? I couldn't see the code that does that. 

Yes, when `SubmitTransaction` is called, the nonce is then incremented when a transaction is being processed here:
https://github.com/spacemeshos/go-spacemesh/blob/develop/pendingtxs/pending_txs.go#L123

In other words, I followed this rule when adding spawn TXs:

1) If a test calls `SubmitTransaction`, then a transaction will be processed and the nonce will be incremented. When `SubmitTransaction` is called, tests start with account nonce == 0. For this case, I added spawn TXs.

2) If a test doesn't call `SubmitTransaction`, a transaction is not processed, so there's no nonce increment. The tests start with nonce != 0 in this case. For this case, I didn't add spawn TXs.

Please let me know if my explanation is not clear.

nkryuchkov commented:
bors merge
